device/cuda/CMakeLists.txt

```diff
@@ -106,7 +106,14 @@ target_compile_options( traccc_cuda
   PRIVATE $<$<COMPILE_LANGUAGE:CUDA>:--expt-relaxed-constexpr> )
 target_link_libraries( traccc_cuda
   PUBLIC traccc::core detray::core vecmem::core covfie::core
-  PRIVATE CUDA::cudart traccc::Thrust traccc::device_common vecmem::cuda )
+  PRIVATE CUDA::cudart traccc::device_common vecmem::cuda )
+
+# Set up Thrust specifically for the traccc::cuda library.
+thrust_create_target( traccc::cuda_thrust
+  HOST CPP
+  DEVICE CUDA )
+target_link_libraries( traccc_cuda
+  PRIVATE traccc::cuda_thrust )
```

Comment:
Out of curiosity, if you link a target against multiple of these Thrust libraries, what happens? Does it give an error at configuration time or does it cause include ambiguities? :thinking:

Reply:
I'm desperately trying to find it in the code, but don't remember all the details anymore. :frowning:

I **think** the issue comes from the specific Thrust targets setting some custom target properties on themselves. See:

https://github.com/NVIDIA/cccl/blob/v2.6.1/thrust/thrust/cmake/thrust-config.cmake#L447-L448

If I remember correctly what I saw was that in case a target ends up being linked against let's say both `Thrust::CPP::Device` and `Thrust::CUDA::Device`, then CMake freaks out during the build system generation stage. Stating that the value of `THRUST_INTERFACE_DEVICE` couldn't be determined for the user's target. Or something like that.

But definitely, already during the CMake configuration one gets an error if trying to link against multiple "host" or "device" libraries at the same time. This is why `traccc::Thrust` is set up in this hacky way now. That I really can't use `thrust_create_target(...)` to make a target that would be appropriate for `traccc::core`. :frowning: